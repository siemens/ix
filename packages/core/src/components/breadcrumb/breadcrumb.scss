/*
 * SPDX-FileCopyrightText: 2023 Siemens AG
 *
 * SPDX-License-Identifier: MIT
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

@import 'common-variables';
@import 'mixins/hover';
@import 'mixins/text-truncation';
@import 'mixins/shadow-dom/component';

@mixin crumb-hover($variant) {
  @include hover {
    background-color: var(--theme-breadcrumb-#{$variant}--background--hover);

    .crumb-text,
    ix-icon {
      color: var(--theme-breadcrumb-#{$variant}--color--hover);
    }

    .crumb-text + ix-icon {
      color: var(--theme-breadcrumb-#{$variant}-arrow--color--hover);
    }
  }

  @include active {
    background-color: var(--theme-breadcrumb-#{$variant}--background--active);

    .crumb-text,
    ix-icon {
      color: var(--theme-breadcrumb-#{$variant}--color--active);
    }

    .crumb-text + ix-icon {
      color: var(--theme-breadcrumb-#{$variant}-arrow--color--active);
    }
  }
}

@mixin breadcrumb($variant) {
  .crumb.#{$variant} {
    background-color: var(--theme-breadcrumb-#{$variant}--background);
    border: var(--theme-btn--#{$variant}-thickness) solid transparent;
    transition: $default-time;

    .crumb-text,
    ix-icon {
      color: var(--theme-breadcrumb-#{$variant}--color);
    }

    .crumb-text + ix-icon {
      color: var(--theme-breadcrumb-#{$variant}-arrow--color);
    }

    &.last .crumb-text {
      color: var(--theme-breadcrumb-#{$variant}-last--color);
    }

    &.clickable:not(.remove-hover) {
      @include hover {
        background-color: var(
          --theme-breadcrumb-#{$variant}--background--hover
        );

        .crumb-text,
        ix-icon {
          color: var(--theme-breadcrumb-#{$variant}--color--hover);
        }

        .crumb-text + ix-icon {
          color: var(--theme-breadcrumb-#{$variant}-arrow--color--hover);
        }
      }

      @include active {
        background-color: var(
          --theme-breadcrumb-#{$variant}--background--active
        );

        .crumb-text,
        ix-icon {
          color: var(--theme-breadcrumb-#{$variant}--color--active);
        }

        .crumb-text + ix-icon {
          color: var(--theme-breadcrumb-#{$variant}-arrow--color--active);
        }
      }
    }
  }
}

:host {
  @include ix-component;

  display: flex;
  height: $large-control-height;
  justify-content: flex-start;
  align-items: center;
  background-color: transparent;
  overflow: hidden;

  .crumb {
    position: relative;
    display: flex;
    justify-content: flex-start;
    align-items: center;
    white-space: nowrap;
    border-radius: var(--theme-breadcrumb--border-radius);
    height: $default-control-height;
    max-height: $default-control-height;
    min-width: 0;
    margin-inline-end: $tiny-space;
    padding: $small-space 0 $small-space $tiny-space;
    outline: none;
    cursor: pointer;

    @include crumb-hover('btn');

    .crumb-text {
      @include ellipsis;
      display: flex;
      align-items: center;
      min-width: 0;
      font-weight: $font-weight-bold;

      ix-icon {
        margin-inline-end: $tiny-space;
      }
    }

    .crumb-text + ix-icon {
      margin-inline-start: $tiny-space;
    }

    &.last {
      background-color: transparent !important;
      cursor: default;
      padding-inline-end: $tiny-space;
    }

    &.clickable:not(.remove-hover) {
      @include focus-visible {
        border-color: var(--theme-focus--border-color);
      }
    }

    &:not(:last-of-type) {
      flex-shrink: 0;
      max-width: 14rem;
    }
  }

  .crumb-dropdown {
    overflow: visible;
  }

  .remove-anchor::after {
    display: none;
  }

  .more-text {
    display: flex;

    .more-text-ellipsis {
      width: 1rem;
      display: inline-block;
      font-weight: $font-weight-bold;
    }

    ix-icon {
      // line-height: unset; --> Note: 'line-height' can't be applied due to shadow DOM encapsulation at ix-icon
      // using padding-top to achieve same position as before
      padding-top: 2px;
    }
  }

  .crumb-items {
    display: contents;
  }

  @include breadcrumb('btn');
  @include breadcrumb('ghost');
}
