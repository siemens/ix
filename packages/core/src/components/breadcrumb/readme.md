<!--
SPDX-FileCopyrightText: 2022 Siemens AG

SPDX-License-Identifier: MIT
-->

# cui-breadcrumb

<!-- Auto Generated Below -->

## Properties

| Property           | Attribute            | Description                                                                                                       | Type       | Default |
| ------------------ | -------------------- | ----------------------------------------------------------------------------------------------------------------- | ---------- | ------- |
| `ghost`            | `ghost`              | Ghost breadcrumbs will not show solid backgrounds on individual crumbs unless there is a mouse event (e.g. hover) | `boolean`  | `false` |
| `nextItems`        | --                   | Items will be accessible through a dropdown                                                                       | `string[]` | `[]`    |
| `visibleItemCount` | `visible-item-count` | Excess items will get hidden inside of dropdown                                                                   | `number`   | `9`     |

## Events

| Event       | Description              | Type                                             |
| ----------- | ------------------------ | ------------------------------------------------ |
| `itemClick` | Crumb item clicked event | `CustomEvent<string>`                            |
| `nextClick` | Next item clicked event  | `CustomEvent<{ event: UIEvent; item: string; }>` |

---
