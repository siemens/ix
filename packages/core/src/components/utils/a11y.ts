/*
 * SPDX-FileCopyrightText: 2023 Siemens AG
 *
 * SPDX-License-Identifier: MIT
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
export const a11yBoolean = (value: boolean) => (value ? 'true' : 'false');

export const a11yHostAttributes = (
  hostElement: HTMLElement
): Record<A11yAttributeName, any> => {
  const attributeObject: Record<string, any> = {};
  a11yAttributes.forEach((attr) => {
    if (hostElement.hasAttribute(attr)) {
      const value = hostElement.getAttribute(attr);
      if (value !== null) {
        attributeObject[attr] = hostElement.getAttribute(attr);
      }
      hostElement.removeAttribute(attr);
    }
  });

  return attributeObject;
};

type PartialRecord<K extends keyof any, T> = {
  [P in K]?: T;
};
export type A11yAttributes = PartialRecord<A11yAttributeName, any>;

export type A11yAttributeName =
  | 'role'
  | 'aria-activedescendant'
  | 'aria-atomic'
  | 'aria-autocomplete'
  | 'aria-braillelabel'
  | 'aria-brailleroledescription'
  | 'aria-busy'
  | 'aria-checked'
  | 'aria-colcount'
  | 'aria-colindex'
  | 'aria-colindextext'
  | 'aria-colspan'
  | 'aria-controls'
  | 'aria-current'
  | 'aria-describedby'
  | 'aria-description'
  | 'aria-details'
  | 'aria-disabled'
  | 'aria-errormessage'
  | 'aria-expanded'
  | 'aria-flowto'
  | 'aria-haspopup'
  | 'aria-hidden'
  | 'aria-invalid'
  | 'aria-keyshortcuts'
  | 'aria-label'
  | 'aria-labelledby'
  | 'aria-level'
  | 'aria-live'
  | 'aria-multiline'
  | 'aria-multiselectable'
  | 'aria-orientation'
  | 'aria-owns'
  | 'aria-placeholder'
  | 'aria-posinset'
  | 'aria-pressed'
  | 'aria-readonly'
  | 'aria-relevant'
  | 'aria-required'
  | 'aria-roledescription'
  | 'aria-rowcount'
  | 'aria-rowindex'
  | 'aria-rowindextext'
  | 'aria-rowspan'
  | 'aria-selected'
  | 'aria-setsize'
  | 'aria-sort'
  | 'aria-valuemax'
  | 'aria-valuemin'
  | 'aria-valuenow'
  | 'aria-valuetext';

const a11yAttributes: A11yAttributeName[] = [
  'role',
  'aria-activedescendant',
  'aria-atomic',
  'aria-autocomplete',
  'aria-braillelabel',
  'aria-brailleroledescription',
  'aria-busy',
  'aria-checked',
  'aria-colcount',
  'aria-colindex',
  'aria-colindextext',
  'aria-colspan',
  'aria-controls',
  'aria-current',
  'aria-describedby',
  'aria-description',
  'aria-details',
  'aria-disabled',
  'aria-errormessage',
  'aria-expanded',
  'aria-flowto',
  'aria-haspopup',
  'aria-hidden',
  'aria-invalid',
  'aria-keyshortcuts',
  'aria-label',
  'aria-labelledby',
  'aria-level',
  'aria-live',
  'aria-multiline',
  'aria-multiselectable',
  'aria-orientation',
  'aria-owns',
  'aria-placeholder',
  'aria-posinset',
  'aria-pressed',
  'aria-readonly',
  'aria-relevant',
  'aria-required',
  'aria-roledescription',
  'aria-rowcount',
  'aria-rowindex',
  'aria-rowindextext',
  'aria-rowspan',
  'aria-selected',
  'aria-setsize',
  'aria-sort',
  'aria-valuemax',
  'aria-valuemin',
  'aria-valuenow',
  'aria-valuetext',
];
