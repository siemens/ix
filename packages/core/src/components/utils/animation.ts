/*
 * SPDX-FileCopyrightText: 2023 Siemens AG
 *
 * SPDX-License-Identifier: MIT
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

export default abstract class Animation {
  static readonly shortTime = 0;
  static readonly defaultTime = 150;
  static readonly mediumTime = 300;
  static readonly slowTime = 500;
  static readonly xSlowTime = 1000;
}
