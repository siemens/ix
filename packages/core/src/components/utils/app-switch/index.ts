/*
 * SPDX-FileCopyrightText: 2023 Siemens AG
 *
 * SPDX-License-Identifier: MIT
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import { AppSwitchConfiguration } from '../application-layout/context';
import { showModal } from '../modal';

export async function showAppSwitch(config: AppSwitchConfiguration) {
  const modal = document.createElement('ix-application-switch-modal');
  modal.config = config;
  const result = await showModal({
    content: modal,
    size: '840',
    closeOnBackdropClick: true,
  });

  const appSwitchElement = result.htmlElement.querySelector(
    'ix-application-switch-modal'
  );

  return (updateAppSwitchConfig: AppSwitchConfiguration) => {
    appSwitchElement.config = updateAppSwitchConfig;
  };
}
