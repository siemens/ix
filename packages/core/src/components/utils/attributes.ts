/*
 * SPDX-FileCopyrightText: 2024 Siemens AG
 *
 * SPDX-License-Identifier: MIT
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

export type Attributes = Record<string, string>;

export const inheritAttributes = (
  element: HTMLElement,
  attributes: string[] = []
) => {
  const attributeObject: Attributes = {};

  attributes.forEach((attr) => {
    if (element.hasAttribute(attr)) {
      const value = element.getAttribute(attr);
      if (value !== null) {
        const attributeValue = element.getAttribute(attr);

        if (attributeValue) {
          attributeObject[attr] = attributeValue;
        }
      }
      element.removeAttribute(attr);
    }
  });

  return attributeObject;
};
