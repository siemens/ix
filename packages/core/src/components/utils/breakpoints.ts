/*
 * SPDX-FileCopyrightText: 2023 Siemens AG
 *
 * SPDX-License-Identifier: MIT
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

// Implementation of scss breakpoints mixins/_break-points.scss
const mediaQueries = {
  sm: '(max-width: 48em)',
  md: '(min-width: 48.0625em) and (max-width: 80em)',
  lg: '(min-width: 80.0625em)',
} as const;

export type Breakpoint = keyof typeof mediaQueries;

export const matchBreakpoint = (breakpoint: Breakpoint) => {
  if (typeof window !== 'undefined' && (window as any).matchMedia) {
    const mediaQuery = mediaQueries[breakpoint];
    return window.matchMedia(mediaQuery).matches;
  }
  return false;
};
