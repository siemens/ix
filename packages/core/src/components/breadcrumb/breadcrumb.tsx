/*
 * SPDX-FileCopyrightText: 2022 Siemens AG
 *
 * SPDX-License-Identifier: MIT
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import {
  Component,
  Element,
  Event,
  EventEmitter,
  h,
  Host,
  Prop,
  State,
} from '@stencil/core';
import animejs from 'animejs';
import { createMutationObserver } from '../utils/mutation-observer';

@Component({
  tag: 'ix-breadcrumb',
  styleUrl: 'breadcrumb.scss',
  scoped: true,
})
export class Breadcrumb {
  /**
   * Excess items will get hidden inside of dropdown
   */
  @Prop() visibleItemCount = 9;

  /**
   * Items will be accessible through a dropdown
   */
  @Prop() nextItems: string[] = [];

  /**
   * Ghost breadcrumbs will not show solid backgrounds on individual crumbs unless there is a mouse event (e.g. hover)
   */
  @Prop() ghost = false;

  /**
   * Crumb item clicked event
   */
  @Event() itemClick: EventEmitter<string>;

  /**
   * Next item clicked event
   */
  @Event() nextClick: EventEmitter<{ event: UIEvent; item: string }>;

  @State() previousButtonRef: HTMLElement;

  @State() nextButtonRef: HTMLElement;

  @Element() hostElement: HTMLIxBreadcrumbElement;

  get breadcrumbItems() {
    return Array.from(this.hostElement.querySelectorAll('ix-breadcrumb-item'));
  }

  get crumbItems() {
    return Array.from(this.hostElement.querySelectorAll('.crumb-items .crumb'));
  }

  @State() items: { label: string; icon?: string }[] = [];

  private mutationObserver: MutationObserver;

  private onItemClick(item: string) {
    this.itemClick.emit(item);
  }

  componentDidLoad() {
    this.mutationObserver = createMutationObserver(() => {
      const updatedItems = this.getItems();
      const update = () => {
        this.items = updatedItems;
      };

      if (updatedItems.length >= this.items.length) {
        update();
      } else if (updatedItems.length < this.items.length) {
        const last = this.crumbItems[this.crumbItems.length - 1] as HTMLElement;
        this.animationFadeOut(last, () => update());
      }
    });

    this.mutationObserver.observe(
      this.hostElement.querySelector('.crumb-items'),
      {
        subtree: true,
        childList: true,
      }
    );
  }

  componentWillLoad() {
    this.items = this.getItems();
  }

  private getItems() {
    return this.breadcrumbItems.map((item) => {
      return { label: item.label, icon: item.icon };
    });
  }

  disconnectedCallback() {
    this.mutationObserver?.disconnect();
  }

  private animationFadeOut(ref: HTMLElement, complete: () => void) {
    animejs({
      targets: ref,
      translateX: '-100%',
      duration: 150,
      opacity: [1, 0],
      easing: 'linear',
      complete: () => complete(),
    });
  }

  private animationFadeIn(ref: HTMLElement) {
    animejs({
      targets: ref,
      duration: 150,
      opacity: [0, 1],
      translateX: ['-100%', '0%'],
      easing: 'linear',
    });
  }

  private handleLastButtonRef(ref: HTMLDivElement, last: boolean) {
    if (last) {
      this.animationFadeIn(ref);
    }
    if (last && this.nextItems?.length) {
      this.nextButtonRef = ref;
    }
  }

  private sliceHiddenItems() {
    let sliceIndex = 0;

    if (this.items.length > this.visibleItemCount) {
      sliceIndex = this.items.length - this.visibleItemCount;
    }

    return this.items.slice(sliceIndex);
  }

  private clickItem(item: string, last: boolean) {
    if (!last) {
      this.onItemClick(item);
    }
  }

  private renderBreadcrumbItems() {
    return this.sliceHiddenItems().map((item, index, array) => {
      const last = index === array.length - 1;

      const isLastItem = last && !this.nextItems?.length;
      return (
        <div
          ref={(ref) => this.handleLastButtonRef(ref, last)}
          data-breadcrumb={index}
          class={{
            crumb: true,
            clickable: true,
            ghost: this.ghost,
            btn: !this.ghost,
            last: isLastItem,
            'remove-hover': isLastItem,
          }}
          onClick={() => this.clickItem(item.label, last)}
          data-testid="item"
        >
          <span class="crumb-text remove-anchor">
            {item.icon ? <ix-icon name={item.icon} size="16"></ix-icon> : ''}
            {item.label}
          </span>
          {!isLastItem ? (
            <span class="glyph glyph-18 glyph-chevron-right-small text-default-text"></span>
          ) : null}
        </div>
      );
    });
  }

  render() {
    return (
      <Host>
        <ix-dropdown
          trigger={
            this.items?.length > this.visibleItemCount
              ? this.previousButtonRef
              : null
          }
        >
          {this.items
            .slice(0, this.items.length - this.visibleItemCount)
            .map((item) => (
              <ix-dropdown-item
                label={item.label}
                onClick={() => this.onItemClick(item.label)}
              ></ix-dropdown-item>
            ))}
        </ix-dropdown>
        {this.items?.length > this.visibleItemCount ? (
          <div
            class="crumb crumb-dropdown"
            ref={(ref) => (this.previousButtonRef = ref)}
          >
            <span class="remove-anchor more-text">
              <span class="more-text-ellipsis">...</span>
              <span class="glyph glyph-16 glyph-chevron-right"></span>
            </span>
          </div>
        ) : null}
        <div class="crumb-items">
          {this.renderBreadcrumbItems()}
          <slot></slot>
        </div>
        <ix-dropdown trigger={this.nextButtonRef}>
          {this.nextItems?.map((item) => (
            <ix-dropdown-item
              label={item}
              onClick={(e) => {
                this.nextClick.emit({
                  event: e,
                  item,
                });
              }}
            ></ix-dropdown-item>
          ))}
        </ix-dropdown>
      </Host>
    );
  }
}
