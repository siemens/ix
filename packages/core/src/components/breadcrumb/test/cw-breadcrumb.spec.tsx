/*
 * SPDX-FileCopyrightText: 2023 Siemens AG
 *
 * SPDX-License-Identifier: MIT
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import { newSpecPage } from '@stencil/core/testing';
import { fireEvent } from '@testing-library/dom';
import { Breadcrumb } from '../breadcrumb';
import { createMutationObserver } from './../../utils/mutation-observer';

jest.mock('./../../utils/mutation-observer');
jest.mock('animejs');

describe('ix-breadcrumb', () => {
  let observerCallback: Function;
  let page: any;

  beforeEach(async () => {
    //@ts-ignore
    createMutationObserver = jest.fn((callback) => {
      observerCallback = callback;
      return {
        observe: jest.fn(),
      };
    });
    page = await newSpecPage({
      components: [Breadcrumb],
      html: `
        <ix-breadcrumb>
          <ix-breadcrumb-item label="Test 1" />
          <ix-breadcrumb-item label="Test 2" />
          <ix-breadcrumb-item label="Test 3" />
          <ix-breadcrumb-item label="Test 4" />
        </ix-breadcrumb>
      `,
    });
  });

  it('renders', async () => {
    observerCallback();

    await page.waitForChanges();
    expect(page.root).toMatchSnapshot();
  });

  it('emits an event when clicked on item', async () => {
    let callbackSpy = jest.fn();
    page.win.addEventListener('itemClick', callbackSpy);

    let item = page.doc
      .querySelector('ix-breadcrumb')
      .shadowRoot.querySelector('[data-testid="item"]');
    fireEvent.click(item);
    await page.waitForChanges();

    expect(callbackSpy).toHaveBeenCalled();
  });
});
