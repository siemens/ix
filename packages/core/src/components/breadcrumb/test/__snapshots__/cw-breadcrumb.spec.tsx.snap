// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ix-breadcrumb renders 1`] = `
<ix-breadcrumb>
  <mock:shadow-root>
    <ix-dropdown></ix-dropdown>
    <div class="crumb-items">
      <div class="btn clickable crumb" data-breadcrumb="0" data-testid="item">
        <span class="crumb-text remove-anchor">
          <span style="white-space: nowrap; overflow: hidden; text-overflow: ellipsis;"></span>
        </span>
        <ix-icon name="chevron-right-small" size="24"></ix-icon>
      </div>
      <div class="btn clickable crumb" data-breadcrumb="1" data-testid="item">
        <span class="crumb-text remove-anchor">
          <span style="white-space: nowrap; overflow: hidden; text-overflow: ellipsis;"></span>
        </span>
        <ix-icon name="chevron-right-small" size="24"></ix-icon>
      </div>
      <div class="btn clickable crumb" data-breadcrumb="2" data-testid="item">
        <span class="crumb-text remove-anchor">
          <span style="white-space: nowrap; overflow: hidden; text-overflow: ellipsis;"></span>
        </span>
        <ix-icon name="chevron-right-small" size="24"></ix-icon>
      </div>
      <div class="btn clickable crumb last remove-hover" data-breadcrumb="3" data-testid="item">
        <span class="crumb-text remove-anchor">
          <span style="white-space: nowrap; overflow: hidden; text-overflow: ellipsis;"></span>
        </span>
      </div>
      <slot></slot>
    </div>
    <ix-dropdown></ix-dropdown>
  </mock:shadow-root>
  <ix-breadcrumb-item label="Test 1">
    <ix-breadcrumb-item label="Test 2">
      <ix-breadcrumb-item label="Test 3">
        <ix-breadcrumb-item label="Test 4"></ix-breadcrumb-item>
      </ix-breadcrumb-item>
    </ix-breadcrumb-item>
  </ix-breadcrumb-item>
</ix-breadcrumb>
`;
